swig files compilation error

#577 fails to compile

https://travis-ci.org/github/UCL/STIR/jobs/697593231#L3130-3142 says
```
In file included from /home/travis/build/UCL/STIR/src/include/stir/make_array.h:209:0,
                 from /home/travis/build/UCL/STIR/src/include/stir/GeometryBlocksOnCylindrical.h:35,
                 from /home/travis/build/UCL/STIR/src/include/stir/ProjDataInfoBlocksOnCylindricalNoArcCorr.h:32,
                 from /home/travis/build/UCL/STIR/src/include/stir/listmode/CListRecordSAFIR.inl:33,
                 from /home/travis/build/UCL/STIR/src/include/stir/listmode/CListRecordSAFIR.h:261,
                 from /home/travis/build/UCL/STIR/src/include/stir/IO/SAFIRCListmodeInputFileFormat.h:47,
                 from /home/travis/build/UCL/STIR/src/IO/IO_registries.cxx:67:
/home/travis/build/UCL/STIR/src/include/stir/make_array.inl: In instantiation of ‘stir::Array<(num_dimensions + 1), T> 
stir::make_array(const stir::Array<num_dimensions, T>&, const stir::Array<num_dimensions, T>&, const stir::Array<num_dimensions, T>&) [with int num_dimensions = 1; T = float]’:

/home/travis/build/UCL/STIR/src/include/stir/GeometryBlocksOnCylindrical.inl:39:7:   required from here

/home/travis/build/UCL/STIR/src/include/stir/make_array.inl:255:113: error: conversion from ‘stir::NumericVectorWithOffset<stir::Array<1, float>, float>’ to non-scalar type ‘const stir::Array<2, float>’ requested

   const Array<num_dimensions+1,T> a = NumericVectorWithOffset<Array<num_dimensions,T>,T>(make_vector(a0, a1, a2));

make[2]: *** [src/swig/CMakeFiles/_stir.dir/__/IO/IO_registries.cxx.o] Error 1
```
_Originally posted by @KrisThielemans in https://github.com/UCL/STIR/pull/577#issuecomment-644230871_

Underlying cause is https://github.com/UCL/STIR/blob/eff69150ec369d0ed7c75bfac5373277d394392b/src/swig/CMakeLists.txt#L30

--- a/examples/src/demo4_obj_fun.cxx
+++ b/examples/src/demo4_obj_fun.cxx
@@ -8,15 +8,15 @@
   All parameters are parsed from a parameter file.
 
   It illustrates
-	- basic class derivation principles
-	- how to use ParsingObject to have automatic capabilities of parsing
-	  parameters files (and interactive questions to the user)
-	- how to initialise and setup a objective function object
+    - basic class derivation principles
+    - how to use stir::ParsingObject to have automatic capabilities of parsing
+      parameters files (and interactive questions to the user)
+    - how to initialise and setup a objective function object
     - how to compute the objective function (log-likelihood) gradient
     - how a basic iterative optimisation works (or may not, if parameters are altered)
 
   Note that the same functionality could be provided without deriving
-  a new class from ParsingObject. One could have a KeyParser object
+  a new class from stir::ParsingObject. One could have a stir::KeyParser object
   in main() and fill it in directly.
 
   See README.txt in the directory where this file is located.

--- a/src/include/stir/KOSMAPOSL/KOSMAPOSLReconstruction.h
+++ b/src/include/stir/KOSMAPOSL/KOSMAPOSLReconstruction.h
@@ -50,10 +50,10 @@ class PoissonLogLikelihoodWithLinearModelForMean;
   \brief A reconstructor class appropriate for emission data
 
   This class implements the iterative algorithm obtained using the Kernel method (KEM) and Hybrid kernel method (HKEM).
-  This implementation corresponds to the one presented by Deidda D et al, ``Hybrid PET-MR list-mode kernelized expectation maximization reconstruction",
+  This implementation corresponds to the one presented by Deidda D et al, "Hybrid PET-MR list-mode kernelized expectation maximization reconstruction",
   Inverse Problems, 2019, DOI: https://doi.org/10.1088/1361-6420/ab013f. However, this allows
   also sinogram-based reconstruction. Furtermore, it is possible to use multiple side information images (they can be anatomical images or images from
-  different modalities). This extension is called multiplexing-HKEM and is described in Deidda et al. ``Multiplexing Kernelized Expectation Maximization
+  different modalities). This extension is called multiplexing-HKEM and is described in Deidda et al. "Multiplexing Kernelized Expectation Maximization
   Reconstruction for PET-MR." IEEE NSS/MIC Proceedings (NSS/MIC) 2018, DOI: https://doi.org/10.1109/NSSMIC.2018.8824312.
   Each voxel value of the image, \f$ \boldsymbol{\lambda}\f$, can be represented as a
   linear combination using the kernel method.  If we have an image with prior information, we can construct for each voxel
@@ -99,12 +99,12 @@ class PoissonLogLikelihoodWithLinearModelForMean;
   KOSMAPOSL Parameters:=
 
   hybrid:=1
-  sigma m:= 1                                ;is the parameter $\f \sigma_{m} \f$;
-  sigma p:=1                                 ;is the parameter $\f \sigma_{p} \f$;
-  sigma dm:=1                                ;is the parameter $\f \sigma_{dm} \f$;
-  sigma dp:=1                                ;is the parameter $\f \sigma_{dp} \f$;
+  sigma m:= 1
+  sigma p:=1 
+  sigma dm:=1
+  sigma dp:=1
   number of neighbours:= 3                   ;is the cubic root of the number of voxels in the neighbourhood;
-  anatomical image filenames:={filename1,filename2}        ;are the filenames of the anatomical images, currently can only do one image;
+  anatomical image filenames:={filename1,filename2}        ;are the filenames of the anatomical images
   number of non-zero feature elements:=1     ;is the number of non zero elements in the feature vector. This makes you choose the size of your feature vector by default we only have one element;
   only_2D:=0                                 ;=1 if you want to reconstruct 2D images;
 
@@ -190,6 +190,7 @@ public:
   void set_sigma_dm(const double);
   void set_only_2D(const bool);
   void set_hybrid(const bool);
+  //@}
 
   //! prompts the user to enter parameter values manually
   virtual void ask_parameters();
@@ -247,7 +248,7 @@ private:
   /*! Estimate the SD of the anatomical image to be used as normalisation for the feature vector */
     void estimate_stand_dev_for_anatomical_image(std::vector<double> &SD);
 
-  /*! Compute for each voxel, jl, of the emission image the linear combination between the coefficient \f$ \alpha_{jl} \f$ and the kernel matrix \f$ k_{jl} \f$\f$ */
+  /*! Compute for each voxel, jl, of the emission image the linear combination between the coefficient \f$ \alpha_{jl} \f$ and the kernel matrix \f$ k_{jl} \f$ */
   /*! The information is stored in the image, kImage */
 //    void full_compute_kernelised_image(TargetT& kernelised_image_out,
 //                                     const TargetT& image_to_kernelise,

--- a/src/include/stir/doxygengroups.h
+++ b/src/include/stir/doxygengroups.h
@@ -209,7 +209,7 @@ from/to file.
 
 /*!
 \defgroup GE support for GE data
-\ingroup GE
+\ingroup IO
 */
 
 /*! 

--- a/src/include/stir/recon_buildblock/PLSPrior.h
+++ b/src/include/stir/recon_buildblock/PLSPrior.h
@@ -50,7 +50,7 @@ START_NAMESPACE_STIR
   https://doi.org/10.1109/TMI.2016.2549601
   Note that PLS becomes smoothed TV when an uniform anatomical image is provided.
 
-  The prior has 2 parameters alpha and eta. It is computed for an image \f$ f f$ as
+  The prior has 2 parameters alpha and eta. It is computed for an image \f$ f \f$ as
 
   \f[
   \phi(f) = \sqrt{\alpha^2 + |\nabla f|^2 - {\langle\nabla f,\xi\rangle}^2}
@@ -62,7 +62,7 @@ START_NAMESPACE_STIR
   \xi = \frac{\nabla v}{\sqrt{|\nabla v|^2 + \eta^2}}
   \f]
 
-  with \f$ v f$ the anatomical image, \f$ \alpha \f$ controls the edge-preservation property of PLS, and depends on the scale
+  with \f$ v \f$ the anatomical image, \f$ \alpha \f$ controls the edge-preservation property of PLS, and depends on the scale
   of the emission image,  and \f$ \eta \f$ avoids division by zero, and depends on the scale of the anatomical image.
 
 

--- a/src/include/stir/recon_buildblock/ProjMatrixByBinSPECTUB.h
+++ b/src/include/stir/recon_buildblock/ProjMatrixByBinSPECTUB.h
@@ -143,7 +143,7 @@ class ProjMatrixByBinSPECTUB :
 
     \verbatim
      sigma_at_depth = collimator_slope * depth_in_mm + collimator sigma 0
-    \end_verbatim
+    \endverbatim
 
     Set slope and sigma_0 to zero for "geometrical" modelling.
 

--- a/src/recon_buildblock/BinNormalisationSPECT.cxx
+++ b/src/recon_buildblock/BinNormalisationSPECT.cxx
@@ -19,7 +19,6 @@
 /*!
   \file
   \ingroup recon_buildblock
-  \ingroup
 
   \brief Implementation for class stir::BinNormalisationSPECT
 

--- a/src/scatter_utilities/upsample_and_fit_single_scatter.cxx
+++ b/src/scatter_utilities/upsample_and_fit_single_scatter.cxx
@@ -47,7 +47,7 @@
   type:= <type info as usual>
   <other keywords for the chosen type>
   END:=
-  \code
+  \endcode
 */
 
 #include "stir/ProjDataInfo.h"

--- a/src/swig/CMakeLists.txt
+++ b/src/swig/CMakeLists.txt
@@ -26,8 +26,6 @@ if(BUILD_SWIG_PYTHON OR BUILD_SWIG_OCTAVE OR BUILD_SWIG_MATLAB)
   SET(CMAKE_SWIG_FLAGS -DSTART_NAMESPACE_STIR=\"namespace stir {\" -DEND_NAMESPACE_STIR=\"}\")
 
   SET_SOURCE_FILES_PROPERTIES(stir.i PROPERTIES CPLUSPLUS ON)
-  # some include files in STIR have "#ifdef SWIG" statements, so we defined the variable here
-  add_definitions(-D SWIG)
 
   if(BUILD_DOCUMENTATION)
     # We use doxy2swig to convert the doxygen-generated xml to a swig file
@@ -69,15 +67,19 @@ if(BUILD_SWIG_PYTHON)
   INCLUDE_DIRECTORIES(${PYTHON_INCLUDE_PATH})
   FIND_PACKAGE(Numpy REQUIRED)
   INCLUDE_DIRECTORIES(${NUMPY_INCLUDE_DIRS})
+  # some include files in STIR have "#ifdef SWIG" statements, so we define the variable here
   # TODO probably better to call the module stirpy or something
   # TODO -builtin option only appropriate for python
   # while the next statement sets it for all modules called stir
-  SET(SWIG_MODULE_stir_EXTRA_FLAGS -builtin ${SWIG_DOXY_OPTIONS})
+  SET(SWIG_MODULE_stir_EXTRA_FLAGS -builtin ${SWIG_DOXY_OPTIONS} -DSWIGXXX)
   if (CMAKE_VERSION VERSION_LESS "3.8")
     SWIG_ADD_MODULE(stir python stir.i ${STIR_REGISTRIES})
   else()
     SWIG_ADD_LIBRARY(stir LANGUAGE python TYPE MODULE SOURCES stir.i ${STIR_REGISTRIES})
   endif()
+  SET_TARGET_PROPERTIES(${SWIG_MODULE_stir_REAL_NAME} PROPERTIES
+            COMPILE_FLAGS -DSWIG # some include files in STIR have "#ifdef SWIG" statements, so we define the variable here
+  )
   SWIG_LINK_LIBRARIES(stir ${STIR_LIBRARIES} ${PYTHON_LIBRARIES})
   target_link_libraries(${SWIG_MODULE_stir_REAL_NAME} ${OpenMP_EXE_LINKER_FLAGS})
   if (SWIG_DOXY_TARGET)
@@ -120,13 +122,16 @@ if (BUILD_SWIG_OCTAVE)
   SET(OCTAVE_SUFFIX ".oct")
   SET(OCTAVE_PREFIX "")
 
-  SET(SWIG_MODULE_stiroct_EXTRA_FLAGS -module stiroct ${SWIG_DOXY_OPTIONS})
+  # some include files in STIR have "#ifdef SWIG" statements, so we define the variable here
+  SET(SWIG_MODULE_stiroct_EXTRA_FLAGS -module stiroct ${SWIG_DOXY_OPTIONS} -DSWIGXXX)
   if (CMAKE_VERSION VERSION_LESS "3.8")
     SWIG_ADD_MODULE(stiroct octave stir.i ${STIR_REGISTRIES})
   else()
     SWIG_ADD_LIBRARY(stiroct LANGUAGE octave TYPE MODULE SOURCES stir.i ${STIR_REGISTRIES})
   endif()
-  SET_TARGET_PROPERTIES(${SWIG_MODULE_stiroct_REAL_NAME} PROPERTIES SUFFIX ${OCTAVE_SUFFIX} PREFIX "${OCTAVE_PREFIX}")
+  SET_TARGET_PROPERTIES(${SWIG_MODULE_stiroct_REAL_NAME} PROPERTIES SUFFIX ${OCTAVE_SUFFIX} PREFIX "${OCTAVE_PREFIX}"
+            COMPILE_FLAGS -DSWIG # some include files in STIR have "#ifdef SWIG" statements, so we define the variable here
+  )
   SWIG_LINK_LIBRARIES(stiroct ${STIR_LIBRARIES} ${OCTAVE_LIBRARIES})
   if (SWIG_DOXY_TARGET)
     add_dependencies(${SWIG_MODULE_stiroct_REAL_NAME} ${SWIG_DOXY_TARGET})
@@ -159,6 +164,7 @@ if (BUILD_SWIG_MATLAB)
   SET_TARGET_PROPERTIES(${SWIG_MODULE_stirMATLAB_REAL_NAME} PROPERTIES
         SUFFIX "_wrap.${Matlab_MEX_EXT}" PREFIX "${MATLAB_PREFIX}"
         LINK_FLAGS "${Matlab_CXXLINKER_FLAGS}"
+        COMPILE_FLAGS -DSWIG # some include files in STIR have "#ifdef SWIG" statements, so we define the variable here
         FOLDER "Matlab")
   target_link_libraries(${SWIG_MODULE_stirMATLAB_REAL_NAME} ${OpenMP_EXE_LINKER_FLAGS})
 	

--- a/src/test/test_Array.cxx
+++ b/src/test/test_Array.cxx
@@ -38,6 +38,7 @@
 #endif
 
 #include "stir/Array.h"
+#include "stir/make_array.h"
 #include "stir/Coordinate2D.h"
 #include "stir/Coordinate3D.h"
 #include "stir/Coordinate4D.h"
@@ -70,6 +71,16 @@ using std::endl;
 
 START_NAMESPACE_STIR
 
+namespace detail {
+
+  static Array<2,float> test_make_array()
+  {
+    return
+      make_array(make_1d_array(1.F,0.F,0.F),
+                 make_1d_array(0.F,1.F,1.F),
+                 make_1d_array(0.F,-2.F,2.F));
+  }
+}
 
 /*!
   \brief Tests Array functionality
@@ -772,6 +783,27 @@ ArrayTests::run_tests()
     run_IO_tests(t1);
   }
 #endif
+  {
+    cerr << "Testing make_array" <<  endl;
+
+    const Array<2,float> arr1 =
+    	make_array(make_1d_array(1.F,0.F,0.F),
+		   make_1d_array(0.F,1.F,1.F),
+		   make_1d_array(0.F,-2.F,2.F));
+
+    const Array<2,float> arr2(
+    	make_array(make_1d_array(1.F,0.F,0.F),
+		   make_1d_array(0.F,1.F,1.F),
+		   make_1d_array(0.F,-2.F,2.F)));
+
+    const Array<2,float> arr3 = detail::test_make_array();
+    const Array<2,float> arr4(detail::test_make_array());
+
+    check_if_equal(arr1[2][1], -2.F, "make_array element comparison");
+    check_if_equal(arr1, arr2, "make_array inline assignment vs constructor");
+    check_if_equal(arr1, arr3, "make_array inline vs function with assignment");
+    check_if_equal(arr1, arr4, "make_array inline constructor from function");
+  }    
 }
 
 END_NAMESPACE_STIR

